text/0002-poet2-consensus.md

```diff
@@ -0,0 +1,566 @@
+>the enclave.
+
+### Block Publishing
+
+To publish a block in the election phase a validator runs the following procedure 
+on the client side:
+
+1. Start the PoET SGX enclave: ENC
+2. Assemble a Proposed Block (outside the enclave)
+3. Call `(waitCertificate, signature) = ENC.createWaitCertificate(blockDigest)`
+4. Wait `WaitCertificate.WaitTime` seconds
+5. Broadcast `(waitCertificate, signature, block, OPK, PPK)` over the
+   Gossip Network.
+
+   Here `block` is the transaction block identified by `blockDigest`.
+
+>Once a `WaitCertificate` is created, the enclave stores the `BlockNumber` and 
+>`WaitCertificate` in a table. The enclave only allows one `WaitCertificate` per
+>`BlockNumber`. Further, it uses the `LastBlockNumber` to ensure a `WaitCertificate` 
+> is created only for future blocks. This prevents an attacker from 'going back 
+>in time' to request a `WaitCertificate` for a past block.
+
+### Block Verification
+Upon receiving a Block from a peer over the Gossip Network, a server side
+validator first the block for eligibility before it can be allowed to
+participate in consensus.
+
+#### Block Eligibility Checks
+1. Verify the PPK and OPK belong to a registered validator by checking the
+   EndPoint registry.
+
+2. Verify the signature is valid using sender's PPK.
+
+3. Verify the PPK was used by sender to commit less than `K` blocks by checking 
+   EndPoint registry (otherwise sender needs to re-sign).
+
+4. Verify the `WaitCertificate.LocalMean` is correct by comparing
+   against `LocalMean` computed locally.
+
+5. Verify the waitCertificate.blockDigest is a valid ECDSA signature of the
+   SHA256 hash of block using OPK.
+
+6. Verify the sender has been winning elections according to the expected
+   distribution (see z-test documentation).
+
+7. Verify the sender signed up at least `c` committed blocks ago, i.e.,
+   respected the `c` block start-up delay.
+
+8. Verify that the block is not an early-arriving block by computing 
+`CC' = CC + WaitCertificate.WaitTime`, then doing `WC >= CC'`, where `CC` is the 
+existing Chain Clock of the fork being extended by the new block.
+
+>Implementation Note1: While checking for early arriving blocks, implementations
+>may choose to add a constant `E`, representing average delay in seconds due to 
+>network latency. The check will then be `WC >= CC' + E`
+
+>Implementation Note2: There may be cases where a block arrives earlier than 
+>its predecessors. In such cases, the block will need to be cached until its
+>predecessors arrive before it can be checked for Eligibility.
+
+A block passing all checks is considered 'Eligible' and proceeds to participate
+in consensus. It is also further broadcast over the Gossip network to
+neighboring peers.
+
+An early arriving block (where `WC < CC'`) is considered 'Ineligible'. The block
+is cached for `CC' - WC` seconds until it becomes 'Eligible'. It is then
+broadcast to neighboring peers over the Gossip Network.

```

Comment:
it will be broadcast assuming that another, better claim is not received before the timer expires.